app/scripts/directives/uiAceYaml.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+(function() {
+  angular.module('openshiftConsole').component('uiAceYaml', {
+    controller: [
+      '$scope',
+      UIAceYAML
+    ],
+    controllerAs: '$ctrl',
+    bindings: {
+      resource: '=',
+      ngRequired: '<?',
+      showFileInput: '<?'
+    },
+    templateUrl: 'views/directives/ui-ace-yaml.html'
+  });
+
+  function UIAceYAML($scope) {
+    var ctrl = this;
+    var aceEditor;
+
+    var parseYAML = function(strict) {
+      // https://github.com/nodeca/js-yaml#safeload-string---options-
+      return jsyaml.safeLoad(ctrl.model, {
+        // If `strict` is false, allow duplicate keys in the YAML for
+        // compability with `oc create` using the js-yaml `json` option. This
+        // also lets us parse JSON as YAML, where duplicate keys are allowed.
+        json: !strict
+      });
+    };
+
+    var clearAnnotations = function() {
+      aceEditor.getSession().clearAnnotations();
+      $scope.$evalAsync(function() {
+        ctrl.annotations = {};
+      });
+    };
+
+    var setAnnotation = function(e, severity) {
+      var session = aceEditor.getSession();
+      var length = session.getLength();
+      var row = _.get(e, 'mark.line', 0);
+      var col = _.get(e, 'mark.column', 0);
+      var message = e.message || 'Could not parse YAML.';
+
+      // If the error line is reported as being after the last line, use the
+      // last line. This can happen when the error is at the very end of the
+      // document and the user doesn't have a final newline.
+      if (row >= length) {
+        row = length - 1;
+      }
+
+      var annotation = {
+        row: row,
+        column: col,
+        text: message,
+        type: severity
+      };
+      session.setAnnotations([ annotation ]);
+
+      $scope.$evalAsync(function() {
+        ctrl.annotations = {};
+        ctrl.annotations[severity] = [ annotation ];
+      });
+    };
+
+    var setValid = function(valid) {
+      $scope.$evalAsync(function() {
+        ctrl.form.$setValidity('yamlValid', valid);
+      });
+    };
+
+    var updated = function() {
+      // Check for errors, then check for warnings.
+      try {
+        ctrl.resource = parseYAML(false);
+        setValid(true);
```

Comment:
does resetting it here cause unnecessary flashing of the disabled state of the Save button, if going from invalid to invalid?  should we wait to reset it until the try and catch blocks for the warnings test?